src/core/DomainDecomposition.cpp

```diff
@@ -0,0 +1,582 @@
+    /* Worst case we need grid - 1 rounds per direction.
+     * This correctly implies that if there is only one node,
+     * no action should be taken. */
+    int rounds_left = grid[0] + grid[1] + grid[2] - 3;
+    for (; rounds_left > 0; rounds_left--) {
+      exchange_neighbors(displaced_parts, diff);
+
+      auto left_over = boost::mpi::all_reduce(m_comm, displaced_parts.size(),
+                                              std::plus<size_t>());
+
+      if (left_over == 0) {
+        break;
+      }
+    }
+  } else {
+    exchange_neighbors(displaced_parts, diff);
+  }
+
+  if (not displaced_parts.empty()) {
+    auto sort_cell = local_cells()[0];
+
+    for (auto &part : displaced_parts) {
+      runtimeErrorMsg() << "Particle " << part.identity()
+                        << " moved more than"
+                           " one local box length in one timestep.";
+      sort_cell->particles().insert(std::move(part));
+
+      diff.emplace_back(ModifiedList{sort_cell->particles()});
+    }
+  }
+}
+void DomainDecomposition::mark_cells() {
+  int cnt_c = 0;
+
+  m_local_cells.clear();
+  m_ghost_cells.clear();
+
+  for (int o = 0; o < ghost_cell_grid[2]; o++)
+    for (int n = 0; n < ghost_cell_grid[1]; n++)
+      for (int m = 0; m < ghost_cell_grid[0]; m++) {
+        if ((m > 0 && m < ghost_cell_grid[0] - 1 && n > 0 &&
+             n < ghost_cell_grid[1] - 1 && o > 0 && o < ghost_cell_grid[2] - 1))
+          m_local_cells.push_back(&cells.at(cnt_c++));
+        else
+          m_ghost_cells.push_back(&cells.at(cnt_c++));
+      }
+}
+void DomainDecomposition::fill_comm_cell_lists(ParticleList **part_lists,
+                                               const Utils::Vector3i &lc,
+                                               const Utils::Vector3i &hc) {
+  for (int o = lc[0]; o <= hc[0]; o++)
+    for (int n = lc[1]; n <= hc[1]; n++)
+      for (int m = lc[2]; m <= hc[2]; m++) {
+        auto const i = Utils::get_linear_index(o, n, m, ghost_cell_grid);
+
+        *part_lists++ = &(cells.at(i).particles());
+      }
+}
+Utils::Vector3d DomainDecomposition::max_range() const {
+  auto dir_max_range = [this](int i) {
+    return std::min(0.5 * m_box.length()[i], m_local_box.length()[i]);
+  };
+
+  return {dir_max_range(0), dir_max_range(1), dir_max_range(2)};
+}
+int DomainDecomposition::calc_processor_min_num_cells() const {
+  /* the minimal number of cells can be lower if there are at least two nodes
+     serving a direction,
+     since this also ensures that the cell size is at most half the box
+     length. However, if there is only one processor for a direction, there
+     have to be at least two cells for this direction. */
+  return boost::accumulate(Utils::Mpi::cart_get<3>(m_comm).dims, 1,
+                           [](int n_cells, int grid) {
+                             return (grid == 1) ? 2 * n_cells : n_cells;
+                           });
+}
+
+void DomainDecomposition::create_cell_grid(double range) {
```

Comment:
It might be worthwhile to pull out the algorithm which calculates the local cell grid as a function of the comm cart shape and the ia range, such that it can be unit tested.

Reply:
I agree that this should be tested, but to do this we first have to find out what the expected behavior is :-) I'm not sure I understand all of the logic...